### PR TITLE
Add doubleclick events

### PR DESCRIPTION
**Pull request by etpinard**
As requested by @cpsievert in https://github.com/ropensci/plotly/pull/416 .

This PR also introduces jasmine tests for select and lasso event data.

Part of https://github.com/plotly/plotly.js/issues/145

### CODE REVIEW

**Comment by ifll** [comment]
Have you considered having a "plotly_pan", "plotly_zoomin", "plotly_zoomout", "plotly_resetaxis", etc... events?

This would help a lot to control the user interaction from the code side

File: test/jasmine/tests/register_test.js [hunk 1]
```diff
@@ -1,6 +1,27 @@
 var Plotly = require('@lib/index');
 
 describe('the register function', function() {
+    'use strict';
+
+    var Plots = Plotly.Plots;
+
+    beforeEach(function() {
+        this.modulesKeys = Object.keys(Plots.modules);
```
Comment (mdtusz) [inline test/jasmine/tests/register_test.js]: Can we avoid using `this` here?

Comment (etpinard) [inline test/jasmine/tests/register_test.js]: Sure. 

You're not a fan of the jasmine `this` ?

Comment (mdtusz) [inline test/jasmine/tests/register_test.js]: Actually on second look, I'm ok with that. Didn't know that it was jasmine convention and that the `this` is cleared between specs.

File: test/jasmine/tests/select_test.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+        var gd;
+        beforeEach(function(done) {
+            gd = createGraphDiv();
+
+            Plotly.plot(gd, mockCopy.data, mockCopy.layout)
+                .then(done);
+        });
+
+        it('should trigger events', function(done) {
```
Comment (mdtusz) [inline test/jasmine/tests/select_test.js]: Maybe add some comments as to the specifics of the test. It's not immediately clear what's being tested.

File: test/jasmine/tests/select_test.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        mouseEvent('mousedown', path[0][0], path[0][1]);
+
+        path.slice(1, len).forEach(function(pt) {
+            mouseEvent('mousemove', pt[0], pt[1]);
+        });
+
+        mouseEvent('mouseup', path[len - 1][0], path[len - 1][1]);
+    }
+
+    function click(x, y) {
```
Comment (mdtusz) [inline test/jasmine/tests/select_test.js]: Is there a reason for using `mousedown` and `mouseup` over just `click`?

Comment (etpinard) [inline test/jasmine/tests/select_test.js]: See https://github.com/plotly/plotly.js/pull/176#discussion_r49474717

But yeah, I should mention why in the code too.

File: test/jasmine/tests/select_test.js (hunk 2 above)
Comment (etpinard) [inline test/jasmine/tests/select_test.js]: sure

File: test/jasmine/tests/click_test.js [hunk 4]
```diff
@@ -9,47 +10,66 @@ var mouseEvent = require('../assets/mouse_event');
...
+        clientX = 351,
+        clientY = 223,
+        gd;
 
-        beforeEach(function(done) {
-            gd = createGraphDiv();
+    function click() {
+        mouseEvent('mousemove', clientX, clientY);
+        mouseEvent('mousedown', clientX, clientY);
+        mouseEvent('mouseup', clientX, clientY);
```
Comment (mdtusz) [inline test/jasmine/tests/click_test.js]: Is this not equivalent to `mouseEvent('click, ...)`?

**Comment by mdtusz** [comment]
Looks good. Nice tests too. Seems like `closeToArray` will get lots of use in the future :beers:

:dancer:

File: src/plots/cartesian/select.js [hunk 5]
```diff
@@ -171,6 +171,8 @@ module.exports = function prepSelect(e, startX, startY, dragOptions, mode) {
                 searchInfo = searchTraces[i];
                 searchInfo.selectPoints(searchInfo, false);
             }
+
+            gd.emit('plotly_deselect', null);
```
Comment (etpinard) [inline src/plots/cartesian/select.js]: fyi @cpsievert

**Comment by cpsievert** [comment]
:boom:

**Pull request merged**